lib/inferno/dsl/suite_endpoint.rb

```diff
@@ -0,0 +1,204 @@
+require 'hanami/controller'
+require_relative '../ext/rack'
+
+module Inferno
+  module DSL
+    # A base class for creating endpoints to test client requests.
+    class SuiteEndpoint < Hanami::Action
+      attr_reader :req, :res
+
+      # @private
+      def self.call(...)
+        new.call(...)
+      end
+
+      # @return [Inferno::Repositories::Requests]
+      def requests_repo
+        @requests_repo ||= Inferno::Repositories::Requests.new
+      end
+
+      # @return [Inferno::Repositories::Results]
+      def results_repo
+        @results_repo ||= Inferno::Repositories::Results.new
+      end
+
+      # @return [Inferno::Repositories::TestRuns]
+      def test_runs_repo
+        @test_runs_repo ||= Inferno::Repositories::TestRuns.new
+      end
+
+      # @return [Inferno::Repositories::Tests]
+      def tests_repo
+        @tests_repo ||= Inferno::Repositories::Tests.new
+      end
+
+      # @private
+      def initialize(config: self.class.config) # rubocop:disable Lint/MissingSuper
+        @config = config
+      end
+
+      # The incoming request as a `Hanami::Action::Request`
+      #
+      # @return [Hanami::Action::Request]
+      #
+      # @example
+      # request.params               # Get url/query params
+      # request.body                 # Get body
+      # request.headers['accept'] # Get Accept header
+      def request
+        req
+      end
+
+      # The response as a `Hanami::Action::Response`. Modify this to build the
+      # response to the incoming request.
+      #
+      # @return [Hanami::Action::Response]
+      #
+      # @example
+      # response.status = 200        # Set the status
+      # response.body = 'Ok'         # Set the body
+      # # Set headers
+      # response.headers.merge!('X-Custom-Header' => 'CUSTOM_HEADER_VALUE')
+      def response
+        res
+      end
+
+      # Override this method to determine a test run's identifier based on an
+      # incoming request.
+      #
+      # @return [String]
+      #
+      # @example
+      # def test_run_identifier
+      #   # Identify the test session of an incoming request based on the bearer
+      #   # token
+      #   request.headers['authorization']&.delete_prefix('Bearer ')
+      # end
+      def test_run_identifier
+        nil
+      end
+
+      # Override this method to build the response.
+      #
+      # @return [Void]
+      #
+      # @example
+      # def make_response
+      #   response.status = 200
+      #   response.body = { abc: 123 }.to_json
+      #   response.format = :json
+      # end
+      def make_response
```

Comment:
I wanted this to be `build_response`, but that is an existing method in `Hanami::Action`. Very open to better name suggestions.